### DISCUSSION TITLE
.NET 8 & RenderModes

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#3461] [General]
I covered this topic in our last Developer Meeting... however as I have gotten further into the OqtaneSSR POC (https://github.com/oqtane/OqtaneSSR) I have uncovered a few more technical challenges which are important for the Oqtane community to understand...

This .NET 8 documentation related to RenderModes is a must-read for Blazor developers (https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#render-mode-propagation). In particular:

Render modes propagate down the component hierarchy. Rules for applying render modes:

- The default render mode is Static.
- The interactive Server (InteractiveServer), WebAssembly (InteractiveWebAssembly), and automatic (InteractiveAuto) render modes can be used from a Static component, including using different render modes for sibling components.
- You can't switch to a different interactive render mode in a child component. For example, a Server component can't be a child of a WebAssembly component.
- Parameters passed to an interactive child component from a Static parent must be JSON serializable. This means that you can't pass render fragments or child content from a Static parent component to an interactive child component.

and I am going to add a few more:

- CascadingParameters cannot be passed to an interactive child component from a Static parent. The CascadingParameter will be null in the child component. The way to handle this is to ensure the CascadingParameter is serializable and pass it explicitly as a standard Parameter.
- DynamicComponent cannot be used to transition from a Static parent to an interactive child. This is because it relies on a Type parameter which is not serializable - it will throw a run-time error. You will need to use other alternatives.

These rules and limitations are important to understand because you need to understand that a Blazor application in .NET 8 has "Rendermode Boundaries". A RenderMode Boundary is the point at which your application transitions from static processing to interactive processing. You are essentially moving from one application processing model to another, and as part of this transition you will need to manage any state which needs to flow from one model to another. This was never an issue with "Interactive Blazor" as the entire application was considered to be interactive - so there were no rendermode boundaries to deal with.

A default Blazor application typically only has a layout and a single routable component (@page) per page. This is a simplistic model which makes it easier to deal with rendermode boudaries. Oqtane has a much more sophisticated UI composition model which allows multiple components per page... as well as a variety of other functionality to consider.

First, Oqtane is multi-tenant - it allows multiple Sites to the created within an installation - each with their own content, etc... Currently Oqtane allows you to specify your Runtime Hosting Model and whether or not Prerendering is enabled. In the SSR version of Oqtane these options will be replaced with InteractiveRenderMode which will allow you to configure an individual Site to use InteractiveServer, InteractiveWebAssembly, or InteractiveAuto as the default interactive render mode for interactive components in the site.

Second, Oqtane is modular. The IModuleControl interface will be enhanced with a new property - RenderMode. The acceptable values will be Interactive or Static (default will be Interactive when not specified so that legacy modules function properly). This will allow a module to tell the framework whether or not it requires interactivity (as the developer of the module knows its rendering mode dependencies). If a module component utilizes multiple rendering modes based on run-time context, it will need to specify Interactive.

When the Composite UI is rendered it will utilize the RenderMode specification for each module component on a page. If the RenderMode is Interactive, it will use the InteractiveRenderMode defined at the Site level to make the component interactive (by setting its @rendermode).

Now let's consider an Oqtane site created from the default template:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/4795ac94-5370-4fe9-bc4b-ddf8d0788f04)

All of the gray outlines represent components which theoretically can be statically rendered (non-interactive)
All of the yellow outlines represent components which currently require interaction
Note that it may be possible to modify some of these interactive components so that they use static techniques instead - it depends on their functionality
It is also important to note that some components contain both interactive and non-interactive requirements and are dependent upon run-time context. For example, if an unauthenticated visitor views the page they will not have access to a variety of the admin components... however once they login, these components will become visible.

I used the phrase "theoretically can be statically rendered" above as this is a critical aspect which needs to be established for Oqtane. To put this into perspective let's consider Theme components... 

Theme components generally contain static content... which make them a perfect candidate for static rendering. And the components used within a Theme would be considered sibling components, so it would be possible for some of them to be static and some of them to be interactive (ie. Logo could be static and Login could be interactive). However if a developer includes interactive functionality directly in their Theme component (which is possible today) it would create some problems. This is because it is the responsibility of the Oqtane router to dynamically construct the UI and decorate components with @rendermode attributes. So if Oqtane only supported static Themes then it would never include an interactive @rendermode attribute - which means that Themes could not be interactive. 

So the question will be, what is the downside of allowing Themes to be interactive? We need to revisit the RenderMode rules to understand this. If a Theme is interactive it means that all child components of the Theme must also be interactive (which is the entire rendering model from theme -> pane -> container -> module). So the entire application would essentially become interactive. In addition, the rendermode boundary would essentially be at the Theme level of the hierarchy - meaning that items like PageState would need to be marshalled across the boundary so that it is available within its interactive child components. Supporting multiple rendermode boundaries will be challenging - especially if we want to preserve backward compatibility as all modules/themes in Oqtane currently rely on CascadingParameters for PageState and ModuleState - so we would need to introduce a "shim" component between each hierarchical component layer to handle the rendermode boundary transition. All of these additional layers could affect performance.

What is the benefit of Themes being static (other than performance). The biggest benefit would be allowing Themes to use standard JavaScript without JS Interop (JS Interop requires interactivity).

RenderMode is one of those topics where the concept is fairly simple to grasp, but the technical restrictions of the underlying technology make it very difficult to implement... so much so that it may require altering your original expectation. I encourage you to experiment with the POC... just be aware that it is a work in progress and is changing constantly as new scenarios are explored. Currently the POC only allows Modules to be interactive.... Themes and Containers are static.

**Comment by sbwalker** [comment]
An example of the render mode challenge described above within the OqtaneSSR POC is the "Home" module. It uses the AddModuleMessage() method from ModuleBase to display a message. This message is displayed in the parent component (DynamicRenderMode). 

A module can be configured to use a Static or Interactive render mode. Therefore, the Close button for the message needs to work in both scenarios. So rather than using the standard Blazor approach (which only works in Interactive render mode):

```
<button type="button" class="btn-close" aria-label="Close" @onclick="DismissMessage"></button>

private void DismissMessage()
{
    _messageContent = "";
    StateHasChanged();
}
```

You need to use a static approach (with data-enhance to enable enhanced forms to prevent a full page refresh):

```
<form method="post" @onsubmit="DismissMessage" @formname="@ModuleState.ModuleType" data-enhance>
    <AntiforgeryToken />
    <button type="submit" class="btn-close" aria-label="Close"></button>
</form>

private void DismissMessage()
{
    _messageContent = "";
}
```

**Comment by sbwalker** [comment]
Another very significant challenge in supporting the new Blazor Server-Side Rendering...

Currently in Oqtane when a request is received from a browser, a PageState object is created for the user session and persisted in memory for the life of that browser session. The PageState object contains all of the configuration necessary to render the UI. As a user navigates to various pages in the site, the PageState is used rather than reloading the information from the API/database. This provides a very high performance user experience.

Blazor Server-Side Rendering is a completely stateless model. This means that PageState cannot be persisted in memory for the user session. So this means that as a user navigates to various pages, the information will need to be reloaded from the API/database. This  has the potential to create a LOT of additional workload on the server when compared to the current stateful model... which could impact performance.

This area will require more research.

**Comment by leigh-pointer** [comment]
Please beware of the Auto Rendering issue https://github.com/dotnet/aspnetcore/issues/52154 "**Auto Render not working properly in .NET 8 and page locks until Web Assembly start**".
The Auto render is not good as the Page locks until the web assembly downloads. So when that bug is fixed we will be able to use the Blazor server circuit fully while the web assembly is downloading.

**Comment by sbwalker** [comment]
Here is another interesting bug I found in .NET 8 Blazor SSR: https://github.com/dotnet/aspnetcore/issues/52360

**Reply by leigh-pointer** [reply]
@sbwalker thats nasty! Although I have got in to the habit of generating  unique prefix string for most of my bootstrap components.  If you have a a collaspe bootstrap component in you module and the have a few instances on a page, when you open one the will all open. drove me crazy.  This #52360 should not be happen.
